crypto/evp/keymgmt_lib.c

```diff
@@ -211,3 +225,259 @@ void *evp_keymgmt_util_fromdata(EVP_PKEY *target, EVP_KEYMGMT *keymgmt,
+{
+    /* Unconstify the provider, because evp_first_name() demands it */
+    OSSL_PROVIDER *prov2 = (OSSL_PROVIDER *)EVP_KEYMGMT_provider(keymgmt2);
+    const char *name2 = evp_first_name(prov2, EVP_KEYMGMT_number(keymgmt2));
+
+    return EVP_KEYMGMT_is_a(keymgmt1, name2);
+}
+
+/*
+ * evp_keymgmt_util_match() adheres to the return values that EVP_PKEY_cmp()
+ * and EVP_PKEY_cmp_parameters() return, i.e.:
+ *
+ *  1   same key
+ *  0   not same key
+ * -1   not same key type
+ * -2   unsupported operation
+ */
+static int implements_match(EVP_KEYMGMT *keymgmt)
+{
+    return keymgmt->match != NULL;
+}
+
+int evp_keymgmt_util_match(EVP_PKEY *pk1, EVP_PKEY *pk2, int selection)
+{
+    EVP_KEYMGMT *impmatch_keymgmt1 = NULL, *impmatch_keymgmt2 = NULL;
+    void *keydata1 = NULL, *keydata2 = NULL;
+
+    find_keymgmt(&impmatch_keymgmt1, &impmatch_keymgmt2, &keydata1, &keydata2,
+                 pk1, pk2, &implements_match);
+
+    /* If we found no suitable keymgmt for either key, support is missing */
+    if (impmatch_keymgmt1 == NULL || impmatch_keymgmt1 == NULL)
+        return -2;
+
+    /*
+     * If we don't have matching keymgmt implementations, we check that they
+     * handle the same key type.
+     * We trust that aliases are properly registered.
+     */
+    if (impmatch_keymgmt1 != impmatch_keymgmt2) {
+        if (!match_type(impmatch_keymgmt1, impmatch_keymgmt2)) {
+            ERR_raise(ERR_LIB_EVP, EVP_R_DIFFERENT_KEY_TYPES);
+            return -1;           /* Not the same type */
+        }
+    }
+
+    /*
+     * If the two keymgmt aren't the same, try to prepare any of the two
+     * EVP_PKEYs so they end up matching.
+     */
+    if (impmatch_keymgmt1 != impmatch_keymgmt2)
+        prepare_binary_op(&impmatch_keymgmt1, &impmatch_keymgmt2,
+                          &keydata1, &keydata2, pk1, pk2);
+
+    /* If we still don't have matching keymgmt implementations, we give up */
+    if (impmatch_keymgmt1 != impmatch_keymgmt2)
+        return -2;
+
+    return evp_keymgmt_match(impmatch_keymgmt1, keydata1, keydata2, selection);
+}
+
+static int implements_copy_and_export(EVP_KEYMGMT *keymgmt)
+{
+    return keymgmt->copy != NULL
+        && (keymgmt->import == NULL || keymgmt->export != NULL);
+}
+
+static void free_keydata_for_copy(EVP_KEYMGMT *keymgmt, void *keydata,
+                                  void *orig_keydata)
+{
+    if (keydata != orig_keydata) {
+        evp_keymgmt_freedata(keymgmt, keydata);
+        EVP_KEYMGMT_free(keymgmt);
+    }
+}
+
+int evp_keymgmt_util_copy(EVP_PKEY *to, EVP_PKEY *from, int selection)
+{
+    EVP_KEYMGMT *impcopy_keymgmt_to = NULL, *impcopy_keymgmt_from = NULL;
+    void *keydata_to = NULL, *keydata_from = NULL;
+
+    /*
+     * Find suitable keymgmt.  We insist that the one we find must not
+     * only implement |copy|, but must implement also |export| and if it
+     * implements |import|.  This should ensure that anything we happen to
```

Comment:
I read this "We insist..." sentence several times and still don't understand it.

Reply:
I can see why...  that comment has gone through so many edits that it ended up not being English any more.  Will fix

Reply:
Slightly rephrased